pkg/framework/test/apiserver.go

```diff
@@ -97,14 +103,47 @@ func (s *APIServer) Start() error {
 	}
 }
 
+func (s *APIServer) ensureInitialized() {
+	if s.PathFinder == nil {
+		s.PathFinder = DefaultBinPathFinder
+	}
+	if s.AddressManager == nil {
+		s.AddressManager = &DefaultAddressManager{}
+	}
+	if s.ProcessStarter == nil {
+		s.ProcessStarter = func(command *exec.Cmd, out, err io.Writer) (SimpleSession, error) {
+			return gexec.Start(command, out, err)
+		}
+	}
+	if s.CertDirManager == nil {
+		s.CertDirManager = NewTempDirManager()
+	}
+	if s.Etcd == nil {
+		s.Etcd = &Etcd{}
+	}
+
+	s.stdOut = gbytes.NewBuffer()
+	s.stdErr = gbytes.NewBuffer()
+}
+
 // Stop stops this process gracefully, waits for its termination, and cleans up the cert directory.
-func (s *APIServer) Stop() {
-	if s.session != nil {
-		s.session.Terminate()
-		s.session.Wait(20 * time.Second)
-		err := s.CertDirManager.Destroy()
-		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+func (s *APIServer) Stop() error {
+	if s.session == nil {
+		return nil
+	}
+
+	s.session.Terminate()
+	// TODO have a better way to handle the timeout of Stop()
+	s.session.Wait(20 * time.Second)
+
+	err := s.Etcd.Stop()
+	if err != nil {
+		return err
 	}
+
+	err = s.CertDirManager.Destroy()
```

Comment:
nit: why save error and then return it?

Reply:
Also, shouldn't we try to remove the certdirmanager even if we failed to stop?

Reply:
Same for etcd of course (which reminds me that there is so much in common ...)

Reply:
What to do when the `Stop()` fails?
My first instinct was: We have no idea in which state the system (the APIServer in that case) is. We don't know, if it is still connected to Etcd, if it has files open in de cert dir, ... so for now we just bail out and let the things in the state they are.
In future we can certainly think about if and how we can be smarter about that, if we can recover from certain errors and retry, ... and all those sort of things.